sklearn/ensemble/_stacking.py

```diff
@@ -0,0 +1,654 @@
+"""Stacking classifier and regressor."""
+
+# Authors: Guillaume Lemaitre <g.lemaitre58@gmail.com>
+# License: BSD 3 clause
+
+from abc import ABCMeta, abstractmethod
+from copy import deepcopy
+
+import numpy as np
+from joblib import Parallel, delayed
+
+from ..base import clone
+from ..base import ClassifierMixin, RegressorMixin, TransformerMixin
+from ..base import is_classifier, is_regressor
+from ..base import MetaEstimatorMixin
+
+from .base import _parallel_fit_estimator
+
+from ..linear_model import LogisticRegression
+from ..linear_model import LinearRegression
+
+from ..model_selection import cross_val_predict
+from ..model_selection import check_cv
+
+from ..preprocessing import LabelEncoder
+
+from ..utils import Bunch
+from ..utils.metaestimators import _BaseComposition
+from ..utils.metaestimators import if_delegate_has_method
+from ..utils.multiclass import check_classification_targets
+from ..utils.validation import check_is_fitted
+from ..utils.validation import column_or_1d
+
+
+class _BaseStacking(_BaseComposition, MetaEstimatorMixin, TransformerMixin,
+                    metaclass=ABCMeta):
+    """Base class for stacking method.
+
+    Warning: This class should not be used directly. Use derived classes
+    instead.
+    """
+    _required_parameters = ['estimators']
+
+    @abstractmethod
+    def __init__(self, estimators, final_estimator=None, cv=None,
+                 stack_method='auto', n_jobs=None, verbose=0):
+        self.estimators = estimators
+        self.final_estimator = final_estimator
+        self.cv = cv
+        self.stack_method = stack_method
+        self.n_jobs = n_jobs
+        self.verbose = verbose
+
+    def _validate_final_estimator(self, default):
+        if self.final_estimator is not None:
+            self.final_estimator_ = clone(self.final_estimator)
+        else:
+            self.final_estimator_ = clone(default)
+
+    @staticmethod
+    def _method_name(name, estimator, method):
```

Comment:
I would argue this applies only to classifiers and should therefore live in the classifier class

Reply:
This is not entirely the case. We also use this code to return `None` if the estimator is `'drop'`. By splitting the validation, we will make spaghetti code. I don't see how to factorize properly the code without creating some `_fit` and several `_validation` functions. Here this is true that we make an additional check for `'predict'` and the regressor but at least fit remain the same.